data/src/main/java/pl/edu/agh/papaya/model/Sprint.java

```diff
@@ -40,4 +43,29 @@
 
     @OneToMany(fetch = FetchType.LAZY, mappedBy = "sprint")
     private List<Notification> notifications;
+
+    public SprintState getSprintState(LocalDateTime currentTime) {
+        if (!closed) {
+            return evaluateNonClosedSprint(currentTime);
+        } else if (currentTime.isBefore(durationPeriod.getEnd())) {
+            /* Editing and closing sprints before their end date is not yet supported,
+               and therefore a specific order needs to be enforced */
+            throw new DateTimeException("A sprint cannot be closed before its end date");
```

Comment:
Nie podoba mi sie ten wyjatek. W dodatku komentarz (część 'editing') wydaje mi sie kompletnie rozbiezny z przeznaczeniem funkcji, kompletnie go nie rozumiem.
Wnioskujac po jego tresci i umiejscowieniu mamy jakby precondition do tej metody pokroju "currentTime nigdy nie jest data z przeszlosci". Wydaje mi sie to dosyc naciągane.
Co chcemy zeby sie stalo gdy podamy gdy podamy currentTime jako date z przeszlosci?, kiedy sprint jeszcze nie byl zamkniety? (moja przychylność w kolejności malejącej)
 - czy chcemy zeby bylo podane ze w tym czasie byl w stanie X (X != CLOSED, bo jeszcze nie byl zamkniety (*)
 - czy chcemy zalozyc ze byl jesli zostal zamkniety to dokladnie w momencie gdy time == durationPeriod.getEnd() i zwrocic CLOSE (to powoduje maly problem z niescisloscia w czasie, bo efektywnie zamykamy go jakby wczesniej niz w rzeczywistosci (patrzac na historie))
 - czy rzucic wyjatkiem w getterze obiektu ktory zawsze ma poprawny stan i jedyny argument zawsze ma sens.
   - efektywnie ten wyjatek w takiej formie jak tutaj powinien moc istniec jedynie przy metodzie zamykajacej sprint

*Jesli chcemy to poprawnie obsluzyc to musimy trzymac date zamkniecia zamiast po prostu flagi. Trzeba minimalnie zmienic model.

Reply:
No wiem, miałam proponować `closed` jako datę

Reply:
A komentarz jak dla mnie ma sens - nie będzie można zamknąć sprintu przed datą zakończenia ani edytować jego dat (w szczególności po zamknięciu). Generalnie będziemy wymuszać pipeline `data1 < data2 < data3 < data4 <  moment_ustawienia_closed_na_true`

Wyjątek wziął się stąd, że to jest niewspierany przez nas stan i nie planujemy do niego dopuszczać, a nie będę zostawiać brancha ifa bez niczego

Reply:
`currentTime` jest (jak wynika z nazwy - choć nie ma tego w nazwie metody) aktualnym czasem. Teraz to zrobiłam tak, że publicznie jest tylko wersja, która w środku robi `LocalDateTime.now()`, a prywatny wariant ma argument z czasem

Reply:
Ale nie twierdzę, że to jest dobre - `closed` miałoby więcej sensu jako data

Reply:
to jeśli to jest _zawsze_ aktualny czas to nie powinno byc tego wyjątku bo kontrakt jest taki, że sprint jest poprawnym obiektem w danym momencie

Reply:
Wariant bez parametru ma na przykład taką cechę, że nie da się go otestować za bardzo (znaczy można jakoś w setupie zamockować zegar pewnie...), a jak czegoś się nie da otestować, to to sugeruje, że może warto przemyśleć to podejście

Reply:
No to dodanie daty zamknięcia zmiana działania tej metody na wyznaczanie stanu w danym momencie w czasie (przeszlosci) rozwiazuje wszystkie nasze problemy - nie ma wyjatku, jest dobrze zdefiniowana, latwa do przetestowania

Reply:
Tak właśnie zrobiłam :) Znaczy to wyznaczanie stanu w przeszłości nie jest naszym bezpośrednim celem, ale sprawia, że ten kod ma więcej sensu